runtime/gc_glue_java/FlattenedArrayObjectScanner.hpp

```diff
@@ -0,0 +1,250 @@
+
+		omrarrayptr_t arrayPtr = (omrarrayptr_t)objectPtr;
+		uintptr_t sizeInElements = arrayObjectModel->getSizeInElements(arrayPtr);
+		uintptr_t elementSize = J9ARRAYCLASS_GET_STRIDE(clazzPtr);
+		env->getExtensions()->mixedObjectModel.getSizeInBytesWithoutHeader(elementClass);
+		fomrobject_t *basePtr = (fomrobject_t *)arrayObjectModel->getDataPointerForContiguous(arrayPtr);
+		fomrobject_t *limitPtr = (fomrobject_t *)((uintptr_t)basePtr + sizeInElements*elementSize);
+		fomrobject_t *scanPtr = (fomrobject_t *)((uintptr_t)basePtr + startIndex*elementSize);
+		fomrobject_t *endPtr = limitPtr;
+		if (!GC_ObjectScanner::isIndexableObjectNoSplit(flags) && (splitAmount != 0)) {
+			Assert_MM_unreachable();
+			endPtr = scanPtr + (splitAmount * elementSize);
+			if (endPtr > limitPtr) {
+				endPtr = limitPtr;
+			}
+		}
+
+		fprintf(stderr, "!!! Flattened array: ");
+		struct J9UTF8* name = J9ROMCLASS_CLASSNAME(elementClass->romClass);
+		fprintf(stderr, "%.*s ", J9UTF8_LENGTH(name), J9UTF8_DATA(name));
+		fprintf(stderr, "elementCount=%ld, ", sizeInElements);
+		fprintf(stderr, "elementSize=%ld, ", elementSize);
+		fprintf(stderr, "(%p, %p)", scanPtr, endPtr);
+		fprintf(stderr, "\n");
+
+		new(objectScanner) GC_FlattenedArrayObjectScanner(env, objectPtr, basePtr, limitPtr, scanPtr, endPtr, 0, elementSize, elementClass, flags);
+		objectScanner->initialize(env, elementClass);
+		if (0 != startIndex) {
+			objectScanner->clearHeadObjectScanner();
+		}
+		return objectScanner;
+	}
+
+	MMINLINE uintptr_t getBytesRemaining() { return sizeof(fomrobject_t) * (_endPtr - _scanPtr); }
+
+	/**
+	 * @param env The scanning thread environment
+	 * @param allocSpace pointer to space within which the scanner should be instantiated (in-place)
+	 * @param splitAmount The maximum number of array elements to include
+	 * @return Pointer to split scanner in allocSpace
+	 */
+	GC_IndexableObjectScanner *
+	splitTo(MM_EnvironmentBase *env, void *allocSpace, uintptr_t splitAmount)
+	{
+		Assert_MM_unimplemented();
+	}
+
+	/**
+	 * Return base pointer and slot bit map for next block of contiguous slots to be scanned. The
+	 * base pointer must be fomrobject_t-aligned. Bits in the bit map are scanned in order of
+	 * increasing significance, and the least significant bit maps to the slot at the returned
+	 * base pointer.
+	 *
+	 * @param[out] scanMap the bit map for the slots contiguous with the returned base pointer
+	 * @param[out] hasNextSlotMap set this to true if this method should be called again, false if this map is known to be last
+	 * @return a pointer to the first slot mapped by the least significant bit of the map, or NULL if no more slots
+	 */
+	virtual fomrobject_t *
+	getNextSlotMap(uintptr_t *slotMap, bool *hasNextSlotMap)
+	{
+		fomrobject_t *result = GC_HeadlessMixedObjectScanner::getNextSlotMap(slotMap, hasNextSlotMap);
+		if (result != NULL) {
+			/* Ignore hasNextSlotMap from HeadLess, we want to always report that there is another element */
```

Comment:
I guess for now. We  might need this for last element of array (as non-mandatory but good to have optimization)

Reply:
sorry, I see how it is implemented. ignore this comment